File: parse.py
import sys
import json
import re

def main():
    with open('storage/app/raw_data.txt', 'r', encoding='utf-8') as f:
        content = f.read()

    blocks = content.split('---')
    
    categories = ['CPU', 'MONITOR', 'PRINTER', 'SWITCH', 'UPS', 'AIO', 'DRIVE']
    
    data = []

    for i, block in enumerate(blocks):
        lines = [line.strip() for line in block.strip().split('\n') if line.strip()]
        if not lines:
            continue
            
        category = categories[i] if i < len(categories) else 'OTHER'
        
        # skip header
        if lines[0].startswith('NO'):
            lines = lines[1:]
            
        for line in lines:
            if category == 'DRIVE':
                # No Merk Type SN Kapasitas
                parts = line.split()
                if len(parts) >= 5:
                    capacity = parts[-1]
                    sn = parts[-2]
                    type_ = parts[-3]
                    merk = " ".join(parts[1:-3])
                    data.append({
                        'brand': merk,
                        'type': type_,
                        'serial_number': sn if sn != '-' else None,
                        'inventory_number': None,
                        'description': f"Kapasitas: {capacity}",
                        'condition': 'Normal',
                        'location': 'GUDANG IT',
                        'category': category
                    })
                continue

            # Default logic for other categories
            parts = line.split()
            if not parts[0].isdigit():
                continue # skip if doesn't start with number
                
            parts = parts[1:] # remove number
            
            # extract condition from the end
            condition = 'Normal'
            last = parts[-1].upper()
            if last in ['RUSAK', 'NORMAL', 'BAIK']:
                if last == 'BAIK':
                    condition = 'Normal'
                else:
                    condition = last.capitalize()
                parts.pop()
                
            # extract location
            location = 'GUDANG IT'
            if len(parts) >= 2 and parts[-2].upper() == 'GUDANG' and parts[-1].upper() == 'IT':
                location = 'GUDANG IT'
                parts.pop()
                parts.pop()

            # extract inventory number
            inv_no = None
            for idx, p in enumerate(parts):
                if p.upper().startswith('IT.'):
                    inv_no = p
                    parts.pop(idx)
                    break
                    
            # determine brand (handle 'TP LINK', 'Western Digital')
            brand = '-'
            if len(parts) > 0:
                if parts[0].upper() == 'TP' and len(parts) > 1 and parts[1].upper() == 'LINK':
                    brand = 'TP-LINK'
                    parts = parts[2:]
                elif parts[0].upper() == 'WESTERN' and len(parts) > 1 and parts[1].upper() == 'DIGITAL':
                    brand = 'Western Digital'
                    parts = parts[2:]
                elif parts[0].upper() == 'ALLIED' and len(parts) > 1 and parts[1].upper() == 'TELESIS':
                    brand = 'ALLIED TELESIS'
                    parts = parts[2:]
                else:
                    brand = parts[0]
                    parts = parts[1:]
                    
            type_ = '-'
            if len(parts) > 0:
                type_ = parts.pop(0)
                
            # try to get SN
            sn = None
            for idx, p in enumerate(parts):
                # SN usually is long and has numbers
                if any(c.isdigit() for c in p) and len(p) >= 5:
                    sn = p
                    parts.pop(idx)
                    break
                    
            description = " ".join(parts) if parts else None
            
            data.append({
                'brand': brand,
                'type': type_,
                'serial_number': sn if sn != '-' else None,
                'inventory_number': inv_no if inv_no != '-' else None,
                'description': description,
                'condition': condition,
                'location': location,
                'category': category
            })
            
    with open('storage/app/parsed_data.json', 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=4)

if __name__ == '__main__':
    main()

File: parse.php
<?php
$content = file_get_contents('storage/app/raw_data.txt');
$blocks = explode('---', $content);
$categories = ['CPU', 'MONITOR', 'PRINTER', 'SWITCH', 'UPS', 'AIO', 'DRIVE'];
$data = [];

foreach ($blocks as $i => $block) {
    $lines = array_filter(array_map('trim', explode("\n", $block)));
    if (empty($lines)) continue;
    $lines = array_values($lines);
    
    $category = $i < count($categories) ? $categories[$i] : 'OTHER';
    
    if (strpos($lines[0], 'NO') === 0 || strpos($lines[0], 'No') === 0) {
        array_shift($lines);
    }
    
    foreach ($lines as $line) {
        $parts = preg_split('/\s+/', $line);
        if (empty($parts) || !is_numeric($parts[0])) continue;
        
        array_shift($parts); // remove number
        
        if ($category === 'DRIVE') {
            if (count($parts) >= 4) {
                $capacity = array_pop($parts);
                $sn = array_pop($parts);
                $type = array_pop($parts);
                $merk = implode(' ', $parts);
                $data[] = [
                    'brand' => $merk,
                    'type' => $type,
                    'serial_number' => $sn !== '-' ? $sn : null,
                    'inventory_number' => null,
                    'description' => "Kapasitas: " . $capacity,
                    'condition' => 'Normal',
                    'location' => 'GUDANG IT',
                    'category' => $category
                ];
            }
            continue;
        }
        
        $condition = 'Normal';
        $last = strtoupper(end($parts));
        if (in_array($last, ['RUSAK', 'NORMAL', 'BAIK'])) {
            if ($last === 'BAIK') $condition = 'Normal';
            else $condition = ucfirst(strtolower($last));
            array_pop($parts);
        }
        
        $location = 'GUDANG IT';
        $count = count($parts);
        if ($count >= 2 && strtoupper($parts[$count-2]) === 'GUDANG' && strtoupper($parts[$count-1]) === 'IT') {
            array_pop($parts);
            array_pop($parts);
        }
        
        $inv_no = null;
        foreach ($parts as $idx => $p) {
            if (strpos(strtoupper($p), 'IT.') === 0) {
                $inv_no = $p;
                unset($parts[$idx]);
                break;
            }
        }
        $parts = array_values($parts);
        
        $brand = '-';
        if (!empty($parts)) {
            if (strtoupper($parts[0]) === 'TP' && count($parts) > 1 && strtoupper($parts[1]) === 'LINK') {
                $brand = 'TP-LINK';
                array_shift($parts); array_shift($parts);
            } elseif (strtoupper($parts[0]) === 'WESTERN' && count($parts) > 1 && strtoupper($parts[1]) === 'DIGITAL') {
                $brand = 'Western Digital';
                array_shift($parts); array_shift($parts);
            } elseif (strtoupper($parts[0]) === 'ALLIED' && count($parts) > 1 && strtoupper($parts[1]) === 'TELESIS') {
                $brand = 'ALLIED TELESIS';
                array_shift($parts); array_shift($parts);
            } else {
                $brand = array_shift($parts);
            }
        }
        
        $type = '-';
        if (!empty($parts)) {
            $type = array_shift($parts);
        }
        
        $sn = null;
        foreach ($parts as $idx => $p) {
            if (preg_match('/\d/', $p) && strlen($p) >= 5) {
                $sn = $p;
                unset($parts[$idx]);
                break;
            }
        }
        $parts = array_values($parts);
        
        $description = !empty($parts) ? implode(' ', $parts) : null;
        
        $data[] = [
            'brand' => $brand,
            'type' => $type,
            'serial_number' => $sn !== '-' ? $sn : null,
            'inventory_number' => $inv_no !== '-' ? $inv_no : null,
            'description' => $description,
            'condition' => $condition,
            'location' => $location,
            'category' => $category
        ];
    }
}

file_put_contents('storage/app/parsed_data.json', json_encode($data, JSON_PRETTY_PRINT));
echo "Parsed " . count($data) . " items.\n";
